### PR TITLE
[MM-45309] Replace github.com/mattermost/gziphandler with github.com/klauspost/compress/gzhttp

### PR DESCRIPTION
**Pull request by hanzei**
#### Summary

The currently used library https://github.com/mattermost/gziphandler is a fork from https://github.com/nytimes/gziphandler, which is unmaintained and dead for 4 years now. There is a long-standing NPE issue that has come up from time to time.

https://pkg.go.dev/github.com/klauspost/compress/gzhttp is a drop-in replacement that is actively maintained. The compression used in `gzhttp` is the same as in https://github.com/mattermost/gziphandler.

One noticeable change with the new library is that some files no longer get compressed (see https://github.com/klauspost/compress/blob/67a538e2b4df11f8ec7139388838a13bce84b5d5/gzhttp/compress.go#L864-L869 for the full list. I _think_ this is actually our wanted behaviour of us as these files are already compressed.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-45309

#### Release Note
```release-note
Replace github.com/mattermost/gziphandler with github.com/klauspost/compress/gzhttp
```

### CODE REVIEW

**Comment by hanzei** [comment]
I would love to measure the potential performance impact of this change. Does the loadtest tool support the static file endpoint? Are the other tools to use?

**Comment by hanzei** [comment]
Todo: check if `gziphandler` can be removed as a dependency.

**Comment by agnivade** [comment]
I see another potential difference. This constant (https://github.com/klauspost/compress/blob/67a538e2b4df11f8ec7139388838a13bce84b5d5/gzhttp/compress.go#L55) is of a different value. We need to run a load test and see if there is any noticeable impact on CPU or response latencies.

> Does the loadtest tool support the static file endpoint?

Good question. I need to check that. I remember us having this discussion a long time back.

> One noticeable change with the new library is that some files no longer get compressed (see https://github.com/klauspost/compress/blob/67a538e2b4df11f8ec7139388838a13bce84b5d5/gzhttp/compress.go#L864-L869 for the full list. I think this is actually our wanted behaviour of us as these files are already compressed.

Hmm .. yes you have a point. cc @streamer45 @agarciamontoro for any thoughts on the matter.

**Comment by hanzei** [comment]
> I see another potential difference. This constant ([klauspost/compress@`67a538e`/gzhttp/compress.go#L55](https://github.com/klauspost/compress/blob/67a538e2b4df11f8ec7139388838a13bce84b5d5/gzhttp/compress.go#L55)) is of a different value. We need to run a load test and see if there is any noticeable impact on CPU or response latencies.

Good catch! I'm not sure if 1024 is a better restriction to use. Using a lower value leaves more "space" for a header, but might lead to unnecessary compression if the original payload would have fit into a MTU anyway.
FYI: I we _really_ want to change that setting, we can do so via the `MinSize()` option.

> > Does the loadtest tool support the static file endpoint?
> 
> Good question. I need to check that. I remember us having this discussion a long time back.

Thanks :+1:

File: server/channels/api4/post.go [hunk 1]
```diff
@@ -651,7 +651,8 @@ func getPostThread(c *Context, w http.ResponseWriter, r *http.Request) {
 	fromPost := r.URL.Query().Get("fromPost")
 	// Either only fromCreateAt must be set, or both fromPost and fromCreateAt must be set
 	if fromPost != "" && fromCreateAt == 0 {
-		c.SetInvalidParam("if fromPost is set, then fromCreatAt must also be set")
+		c.SetInvalidParam("if fromPost is set, then fromCreateAt must also be set")
+		return
```
Comment (hanzei) [inline server/channels/api4/post.go:655]: Previously, if this case was hit, the HTTP response was (correctly) a 400. After the library migration to `github.com/klauspost/compress/gzhttp`, the response is (incorrectly) a 200.

Without this fix, the changes in the PR broke a test. I was not able to track down why the behaviour changed. But the code here was surely wrong and hence I fixed it has part of this PR.

Comment (agnivade) [reply server/channels/api4/post.go:655]: You are the reviewer now: https://github.com/mattermost/mattermost/pull/24288 :)

**Comment by agarciamontoro** [comment]
I don't have an informed opinion on the changes here, but would love to see a load-test comparing the before and after, since this affects to every request. I don't expect much difference, but since this affects to every request, we should at least check.

> Does the loadtest tool support the static file endpoint?

Do you mean the `/api/v4/files/{file_id}` endpoint? If that's what you mean, I don't think we stress that at all. We do upload files, get file previews and thumbnails, but don't download whole files.

**Comment by hanzei** [comment]
Load-test results: https://community.mattermost.com/core/pl/a5fuqemamfyr7b38y4nuqc9rkc

**Review by agarciamontoro — APPROVED** [review]
Neat! Thanks for the thorough analysis :)

**Review by agnivade — APPROVED** [review]
Thanks Ben!

**Pull request merged**